src/Subscription/Engine/DefaultSubscriptionEngine.php

```diff
@@ -962,32 +988,109 @@ private function lowestSubscriptionPosition(array $subscriptions): int
         return $min;
     }
 
-    /**
-     * @param Closure(list<Subscription>):T $closure
-     *
-     * @return T
-     *
-     * @template T
-     */
-    private function findForUpdate(SubscriptionCriteria $criteria, Closure $closure): mixed
+    private function handleError(Subscription $subscription, Throwable $throwable): void
     {
-        if (!$this->subscriptionStore instanceof LockableSubscriptionStore) {
-            return $closure($this->subscriptionStore->find($criteria));
+        $subscription->error($throwable);
+        $this->subscriptionManager->update($subscription);
+
+        if (!isset($this->batching[$subscription->id()])) {
+            return;
         }
 
-        return $this->subscriptionStore->inLock(
-        /** @return T */
-            function () use ($closure, $criteria): mixed {
-                $subscriptions = $this->subscriptionStore->find($criteria);
+        $subscriber = $this->batching[$subscription->id()];
 
-                return $closure($subscriptions);
-            },
-        );
+        unset($this->batching[$subscription->id()]);
+
+        try {
+            $subscriber->rollbackBatch();
+        } catch (Throwable $e) {
+            $this->logger?->error(sprintf(
+                'Subscription Engine: Subscriber "%s" has an error in the rollback batch method: %s',
+                $subscription->id(),
+                $e->getMessage(),
+            ));
+        }
     }
 
-    private function handleError(Subscription $subscription, Throwable $throwable): void
+    private function ensureCommitBatch(Subscription $subscription, int $index): Error|null
     {
-        $subscription->error($throwable);
-        $this->subscriptionStore->update($subscription);
+        if (!isset($this->batching[$subscription->id()])) {
+            return null;
+        }
+
+        $subscriber = $this->batching[$subscription->id()];
+
+        unset($this->batching[$subscription->id()]);
+
+        try {
+            $subscriber->commitBatch();
+            $subscription->changePosition($index);
+        } catch (Throwable $e) {
+            $this->logger?->error(sprintf(
+                'Subscription Engine: Subscriber "%s" has an error in the commit batch method: %s',
+                $subscription->id(),
+                $e->getMessage(),
+            ));
+
+            $this->handleError($subscription, $e);
+
+            return new Error(
+                $subscription->id(),
+                $e->getMessage(),
+                $e,
+            );
+        }
+
+        return null;
+    }
+
+    private function checkAndBeginBatch(Subscription $subscription): Error|null
```

Comment:
Can't we place all the private batching methods into a internal class like you did with the `SubscriptionManager` and just init it in the constructor, not as a dep coming from outside?

Reply:
This is not so easy to do, because otherwise I would have bidirectional dependencies between two classes. I want to avoid that.